src/pack.c

```diff
@@ -965,62 +993,75 @@ int get_delta_base(
  *
  ***********************************************************/
 
-void git_packfile_close(struct git_pack_file *p, bool unlink_packfile)
+/* Run with the mwindow lock held */
+void git_packfile_close_locked(struct git_pack_file *p, bool unlink_packfile)
 {
+	bool locked = true;
+	if (git_mutex_lock(&p->lock) < 0) {
+		git_error_set(GIT_ERROR_OS, "failed to lock pack file");
+		locked = false;
+	}
+	if (locked && git_mutex_lock(&p->mwf.lock) < 0) {
+		git_error_set(GIT_ERROR_OS, "unable to lock packfile");
+		git_mutex_unlock(&p->lock);
+		locked = false;
+	}
 	if (p->mwf.fd >= 0) {
 		git_mwindow_free_all_locked(&p->mwf);
 		p_close(p->mwf.fd);
 		p->mwf.fd = -1;
 	}
+	if (locked) {
+		git_mutex_unlock(&p->mwf.lock);
+		git_mutex_unlock(&p->lock);
+	}
 
 	if (unlink_packfile)
 		p_unlink(p->pack_name);
 }
 
-void git_packfile_free(struct git_pack_file *p)
+/* Run with the mwindow lock held */
+void git_packfile_free_locked(struct git_pack_file *p)
 {
 	if (!p)
 		return;
 
 	cache_free(&p->bases);
 
-	git_packfile_close(p, false);
+	git_packfile_close_locked(p, false);
 
 	pack_index_free(p);
 
 	git__free(p->bad_object_sha1);
 
-	git_mutex_free(&p->lock);
 	git_mutex_free(&p->bases.lock);
+	git_mutex_free(&p->mwf.lock);
+	git_mutex_free(&p->lock);
 	git__free(p);
 }
 
-static int packfile_open(struct git_pack_file *p)
+static int packfile_open_locked(struct git_pack_file *p)
 {
 	struct stat st;
 	struct git_pack_header hdr;
 	git_oid sha1;
 	unsigned char *idx_sha1;
 
-	if (p->index_version == -1 && pack_index_open(p) < 0)
+	if (pack_index_open_locked(p) < 0)
 		return git_odb__error_notfound("failed to open packfile", NULL, 0);
 
-	/* if mwf opened by another thread, return now */
-	if (git_mutex_lock(&p->lock) < 0)
-		return packfile_error("failed to get lock for open");
-
-	if (p->mwf.fd >= 0) {
-		git_mutex_unlock(&p->lock);
+	if (p->mwf.fd >= 0)
 		return 0;
-	}
 
 	/* TODO: open with noatime */
 	p->mwf.fd = git_futils_open_ro(p->pack_name);
 	if (p->mwf.fd < 0)
 		goto cleanup;
 
-	if (p_fstat(p->mwf.fd, &st) < 0 ||
-		git_mwindow_file_register(&p->mwf) < 0)
+	if (p_fstat(p->mwf.fd, &st) < 0)
+		goto cleanup;
```

Comment:
I realize that you didn't introduce this problem, but while you're here, maybe we could:

```suggestion
	if (p_fstat(p->mwf.fd, &st) < 0) {
		git_error_set(GIT_ERROR_OS, "could not stat packfile");
		goto cleanup;
	}
```